main/src/library/Fixpoint/Ast/Datalog.flix

```diff
@@ -66,32 +66,29 @@ mod Fixpoint.Ast.Datalog {
     instance ToString[Datalog[v]] {
         pub def toString(cs: Datalog[v]): String = match cs {
             case Datalog.Datalog(facts, rules) => region rc {
-                let sb = StringBuilder.empty(rc);
-                let appendConstraint = f -> StringBuilder.appendLineWith!(ToString.toString, f, sb);
-                foreach (f <- facts) appendConstraint(f);
-                foreach (ru <- rules) appendConstraint(ru);
-                StringBuilder.toString(sb)
+                Iterator.append(Vector.iterator(rc, facts), Vector.iterator(rc, rules))
+                    |> Iterator.join(String.lineSeparator())
             }
             case Datalog.Model(db) => region rc {
                 use Fixpoint.Ast.Ram.toDenotation;
-                let sb = StringBuilder.empty(rc);
-                Map.forEach(ramSym -> rel -> match toDenotation(ramSym) {
-                    case Denotation.Relational => Map.forEach((tuple, _) -> {
-                        let tupleString = tuple |> Vector.map(Debug.stringify) |> Vector.join(", ");
-                        StringBuilder.appendLine!("${ramSym}(${tupleString}).", sb)
-                    }, rel)
-                    case Denotation.Latticenal(_) => Map.forEach((tuple, lat) -> {
-                        let tupleString = tuple |> Vector.map(Debug.stringify) |> Vector.join(", ");
-                        StringBuilder.appendLine!("${ramSym}(${tupleString}; %{lat}).", sb)
-                    }, rel)
-                }, db);
-                StringBuilder.toString(sb)
+                Map.iterator(rc, db)
+                    |> Iterator.flatMap(match (ramSym, rel) -> {
+                        Map.iterator(rc, rel) |> Iterator.map(match (tuple, lat) -> {
+                            let tupleString = tuple |> Vector.map(Debug.stringify) |> Vector.join(", ");
+                            match toDenotation(ramSym) {
+                                case Denotation.Relational => "${ramSym}(${tupleString})."
+                                case Denotation.Latticenal(_) => "${ramSym}(${tupleString}; %{lat})."
+                            }
+                        })
+                    })
+                    |> Iterator.join(String.lineSeparator())
+                    + String.lineSeparator()
```

Comment:
How confident are we that the above is correct?

Reply:
Did you test it?

Reply:
No. Is there a test in the suite? Should I introduce one?

Reply:
I don't know, but you are welcome to introduce a test.

Reply:
Where would such a test belong? This is part of the Datalog engine, right? How do you obtain a `Datalog.Model` in the Flix world?

Reply:
Yes that is tricky. Lets not do it then.

Reply:
I will ask you to just take one more careful look over the code, and if you feel confident, then we can merge this PR without the test.

Reply:
I'm as confident as I can get without testing it.